overlord/snapstate/handlers.go

```diff
@@ -1934,6 +1934,89 @@ func (m *SnapManager) doUnlinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	return err
 }
 
+func (m *SnapManager) undoUnlinkSnap(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	snapsup, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+
+	// XXX: is it safe (can panic)?
+	isInstalled := snapst.IsInstalled()
+	if !isInstalled {
+		return fmt.Errorf("internal error: snap %q not installed", snapsup.InstanceName())
+	}
+
+	info, err := snapst.CurrentInfo()
+	if err != nil {
+		return err
+	}
+
+	deviceCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+
+	var broken bool
+	// undo here may be part of failed snap remove change, in which case a later
+	// "clear-snap" task could have been executed and some or all of the
+	// data of this snap could be lost. If that's the case, then we should not
+	// enable the snap back.
+	// XXX: should make an exception for snapd/core?
+	chg := t.Change()
+	if chg != nil {
```

Comment:
I'm really not a fan of this logic, can't we detect this by looking at something on disk instead?

Reply:
Yeah, I suppose we could, will check.

Reply:
Done.